File: 169-MajorityElement/169-MajorityElement.cpp
// Last updated: 4/15/2026, 10:48:17 PM
class Solution {
public:
    int majorityElement(vector<int>& nums) {
    int n = nums.size();
       /* Brute force method:
       for(int i=0;i<n;i++){     
         int count=0;
            for(int j=0;j<n;j++){
                if(nums[i]==nums[j]){
                    count++;
                }
            }
            if(count>n/2){
                return nums[i];
            }
        } */


       /*Optimise method(faster and better)
        sort(nums.begin(),nums.end());
        int freq =1;
        int ans = nums[0];
        for(int i=1;i<n;i++){
        if(nums[i]==nums[i-1]){
            freq++;
        }else{
            freq=1;
            ans=nums[i];
        }
        if(freq>n/2){
            return ans;
        }
     } */
    //Moore's Voting Algorithmm:
       int freq=0;
       int ans=0;
       for(int i=0;i<n;i++){
        if(freq==0){
            ans=nums[i];}
        if(ans==nums[i])
           freq++;
        else{
             freq--;}
       }
        return ans;
    }
};

File: 169-MajorityElement/169-MajorityElement.java
// Last updated: 6/28/2026, 10:30:52 PM
1class Solution {
2    public int majorityElement(int[] nums) {
3        //better approach with sorting  T.C. = O(nlogn)
4         Arrays.sort(nums);
5         int n = nums.length;
6         int freq = 1;
7         int ans = nums[0];
8         for(int i =1;i<n;i++){
9          if(nums[i]== nums[i-1]){
10            freq++;
11          }else{
12            freq=1;
13            ans=nums[i];
14          }
15          if(freq > n/2){
16            return ans;
17          }
18         }
19         return ans;
20    }
21}